### PR TITLE
Blog posts integration

### PR DESCRIPTION
**Pull request by nabilwinfotech**
Add Blog posts

- Create a blog route to list out the post.
- Add pagination on that.
- Add search functionality on posts.
- make route and page for individual posts.

Decap CMS Integration

- Integrate Decap cms for all blog posts.
- Create, read, update, delete the posts with decap using `/admin` route.
- All changes in content will reflects on view(UI) and each change make a commit on github repository.

### CODE REVIEW

File: src/admin/config.yml [hunk 1]
```diff
@@ -0,0 +1,39 @@
...
+  delete: Delete {{collection}} “{{slug}}”
+  uploadMedia: Upload “{{path}}”
+  deleteMedia: Delete “{{path}}”
+  openAuthoring: '{{message}}'
+
+
+media_folder: "src/assets/uploads"
+public_folder: "/assets/uploads"
+
+site_url: https://taakconsult.netlify.app/
```
Comment (chiragpaliwal) [inline src/admin/config.yml]: This should be https://taakconsulting.net/

Comment (nabilwinfotech) [reply src/admin/config.yml]: okay, let me change this

**Review by chiragpaliwal — COMMENTED** [review]
We also do not need any placeholder images under `assets/img`

File: src/filters/searchFilter.js [hunk 2]
```diff
@@ -0,0 +1,19 @@
+const elasticlunr = require("elasticlunr");
```
Comment (chiragpaliwal) [inline src/filters/searchFilter.js]: What does this library do?

**Comment by nabilwinfotech** [comment]
> We also do not need any placeholder images under `assets/img`

okay

File: src/posts/post1-title.md
Comment (chiragpaliwal) [inline src/posts/post1-title.md:1]: If we commit these posts, we would end up with dummy posts on our live website which is something we do not intend.

File: src/posts/post1-title.md [hunk 3]
```diff
@@ -0,0 +1,9 @@
+---
+thumbnail: /assets/uploads/screenshot-2024-04-24-100206.png
```
Comment (chiragpaliwal) [inline src/posts/post1-title.md]: How would this thumbnail be picked up for an actual user publishing the blog?

File: src/blog.html [hunk 4]
```diff
@@ -0,0 +1,75 @@
+---
+layout: blog.njk
+pagination:
+  data: collections.post
+  size: 10  # Number of posts per page
+  alias: posts
+  generatePageOnEmptyData: true
+  reverse: true
```
Comment (chiragpaliwal) [inline src/blog.html:8]: What does reverse do?

**Review by chiragpaliwal — CHANGES_REQUESTED** [review]

File: src/filters/searchFilter.js (hunk 2 above)
Comment (nabilwinfotech) [reply src/filters/searchFilter.js]: this is for searching in our blog posts.

File: src/posts/post1-title.md
Comment (nabilwinfotech) [reply src/posts/post1-title.md:1]: okay i delete them and we'll make posts from our decap cms.

File: src/posts/post1-title.md (hunk 3 above)
Comment (nabilwinfotech) [reply src/posts/post1-title.md]: it is all the working of decap, i have generated this with decap and configured all these settings in config.yml that how and what should it do and its responsibility to  manage them

File: src/blog.html (hunk 4 above)
Comment (nabilwinfotech) [reply src/blog.html:8]: it will show latest post on top

**Comment by piyushpaliwal** [comment]
Styling issue on page numbers. Also, pagination doesn't work.

image

**Comment by piyushpaliwal** [comment]
Hovering over page number 2 shows the url as page 1

image

File: src/_includes/blog.njk [hunk 5]
```diff
@@ -0,0 +1,156 @@
...
+                                    name="search-posts"
+                                    id="search-posts"
+                                    type="search"/>
+                            </div>
+                        <ul id="searchResults" class="simple-post-list m-0"></ul>
+                        <div id="noResultsFound" style="display: none">
+                            <p style="color: #777">No results found.</p>
+                        </div>
+                        <script type="text/javascript" src="https://cdnjs.cloudflare.com/ajax/libs/elasticlunr/0.9.6/elasticlunr.min.js"></script>
+                        <script type="text/javascript" src="/js/search.js"></script>
```
Comment (piyushpaliwal) [inline src/_includes/blog.njk]: Script tags in the middle of page body. This should not happen. Script tags should be loaded towards the end of body tag if they are not critical and we should `defer` their loading

**Comment by piyushpaliwal** [comment]
Deploy preview of this PR is full of CORS errors, please investigate what are they, should we take any action or they are just some netlify preview related scripts which are failing.

image

File: src/filters/searchFilter.js (hunk 2 above)
Comment (piyushpaliwal) [reply src/filters/searchFilter.js]: This goes against the quality parameters that we have established. The last commit for this library was 5 years ago and the project is abandoned.

We have talked about library selection criterias numerous times over the past few months.

**Comment by nabilwinfotech** [comment]
> Styling issue on page numbers. Also, pagination doesn't work.
> 
> image

done! and pagination is working

**Comment by nabilwinfotech** [comment]
> Hovering over page number 2 shows the url as page 1
> 
> image

done!

**Comment by nabilwinfotech** [comment]
> Deploy preview of this PR is full of CORS errors, please investigate what are they, should we take any action or they are just some netlify preview related scripts which are failing.
> 
> image

they are  netlify preview related scripts.

File: src/_includes/blog.njk (hunk 5 above)
Comment (nabilwinfotech) [reply src/_includes/blog.njk]: done!

File: src/filters/searchFilter.js (hunk 2 above)
Comment (nabilwinfotech) [reply src/filters/searchFilter.js]: let me resolve this!

File: src/assets/uploads/screenshot-2024-04-24-100206.png
Comment (chiragpaliwal) [inline src/assets/uploads/screenshot-2024-04-24-100206.png:1]: This should be removed.

File: src/_data/nav.json [hunk 6]
```diff
@@ -40,5 +40,10 @@
     "id": 3,
     "title": "About",
     "url": "/about/"
+  },
+  {
+    "id": 4,
+    "title": "Blog",
+    "url": "/blog/1/"
```
Comment (chiragpaliwal) [inline src/_data/nav.json:47]: Let's try to begin the pagination from 2, and base URL can just be `/blog/`

File: src/admin/config.yml [hunk 7]
```diff
@@ -0,0 +1,39 @@
+backend:
+  name: git-gateway
+  branch: decap-integration
```
Comment (chiragpaliwal) [inline src/admin/config.yml]: This should be main before this is pushed to prod.

**Review by chiragpaliwal — APPROVED** [review]
Approving with minor fixes to be addressed later.

- Fix pagination links
- Author hyperlink on article
- Remove any sample / test deployments from personal netlify

@nabilwinfo

**Pull request merged**